In [1]:
import torch
import gc
torch.cuda.empty_cache()
gc.collect()


0

In [2]:
from setup import setup_src_path
print(setup_src_path())
import data.processed as processed
import config.config as config
import utils.setup as setup
import utils.functions as fn
from importlib import reload



['/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/mlm', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '', '/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages', '/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/modules']


2024-06-05 04:22:32.523183: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-06-05 04:22:32.719855: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-06-05 04:22:33.592577: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


In [3]:
from datasets import load_from_disk,concatenate_datasets


adapter_name="mlm_union_all_TE"
dataset = load_from_disk(f"{config.Config.DATASETS_SAVE_PATH}/datasets")



In [4]:
reload(config)

    
filtered_target = dataset['train'].filter(lambda example: example['genre'] == "telephone").shuffle(seed=42)
#filtered_target = shuffled_filtered_target.train_test_split(test_size=0.1)


filtered_test_target = dataset['validation_matched'].filter(lambda example: example['genre'] == "telephone")
train_target = filtered_target
test_target = filtered_test_target
train_target[2]
 

{'promptID': 8103,
 'pairID': '8103e',
 'premise': "but i like dogs and my husband like cats so we haven't reached a real agreement on that yet  if we get a place where we can have both it'll be great but until then",
 'premise_binary_parse': "( but ( ( i ( like ( ( dogs and ) ( ( my husband ) ( like cats ) ) ) ) ) ( so ( we ( ( have n't ) ( reached ( ( a ( real agreement ) ) ( ( on that ) ( ( yet ( if ( we ( get ( ( a place ) ( where ( we ( can ( have both ) ) ) ) ) ) ) ) ) ( it ( 'll ( be ( great ( but ( until then ) ) ) ) ) ) ) ) ) ) ) ) ) ) )",
 'premise_parse': "(ROOT (FRAG (CC but) (ADVP (RB i) (PP (IN like) (NP (NP (NNS dogs)) (CC and) (NP (NP (PRP$ my) (NN husband)) (PP (IN like) (NP (NNS cats))))))) (SBAR (IN so) (S (NP (PRP we)) (VP (VBP have) (RB n't) (VP (VBN reached) (NP (NP (DT a) (JJ real) (NN agreement)) (SBAR (WHPP (IN on) (WHNP (WDT that))) (S (SBAR (RB yet) (IN if) (S (NP (PRP we)) (VP (VBP get) (NP (NP (DT a) (NN place)) (SBAR (WHADVP (WRB where)) (S (NP (PRP we)) (

In [5]:
from adapters import AutoAdapterModel,init
from tqdm import tqdm

from transformers import AutoConfig, AutoModelForMaskedLM

mdlcfg = AutoConfig.from_pretrained(
    config.Config.MODEL_NAME,
 
)
model = AutoModelForMaskedLM.from_pretrained(
    config.Config.MODEL_NAME,
)
init(model)
reload(fn)
fn.print_trainable_parameters(model)

/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertFor

trainable params: 109514298 || all params: 109514298 || trainable%: 100.0


In [6]:
from adapters import ConfigUnion, LoRAConfig, PrefixTuningConfig, SeqBnConfig,SeqBnInvConfig,AdapterConfig,LoRAConfig

union_config = ConfigUnion(
    PrefixTuningConfig(prefix_length=10),
    LoRAConfig(r=8, alpha=16),

    SeqBnInvConfig(reduction_factor=2),
)
model.add_adapter(adapter_name, config=union_config)

model.train_adapter([adapter_name])
model.active_adapters = adapter_name
print(model.adapter_summary())
fn.print_trainable_parameters(model)

Name                     Architecture         #Param      %Param  Active   Train
--------------------------------------------------------------------------------
mlm_union_all_TE         union            17,539,712      16.107       1       1
--------------------------------------------------------------------------------
Full model                               108,891,648     100.000               0
trainable params: 18131840 || all params: 127054010 || trainable%: 14.2709702747674


In [7]:
from transformers import AutoTokenizer

reload(processed)
tokenizer = AutoTokenizer.from_pretrained(config.Config.TOKENIZER_NAME)

tokenized_dataset= processed.tokenize_dataset(train_target,tokenizer)
tokenized_dataset

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids'],
    num_rows: 83348
})

In [8]:
chunk_size = 128
# Slicing produces a list of lists for each feature
tokenized_samples = tokenized_dataset[444:470]

results = fn.group_texts(tokenized_samples, chunk_size)
for chunk in results["labels"]:
    print(f"'>>> Chunk length: {len(chunk)}'")

'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'


In [9]:
# we group texts and chunk them
lm_datasets = tokenized_dataset.map(fn.group_texts, batched=True,fn_kwargs={'chunk_size': chunk_size})
lm_datasets

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
    num_rows: 27551
})

In [10]:
from transformers import DataCollatorForLanguageModeling

data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm_probability=0.15)
samples = [lm_datasets[i] for i in range(2)]
for sample in samples:
    _ = sample.pop("word_ids")
dd=data_collator(samples)
for chunk in dd["input_ids"]:
    print(f"\n'>>> {tokenizer.decode(chunk)}'")


'>>> [CLS] i mean i was absolutely [MASK]abbergasted [MASK] what she [MASK] would that she was paid true [MASK] was [MASK] you know but [MASK] it's the principle of course if she got her master's degree it'd all be different but uh [SEP] i [MASK] she [MASK] much more money. [SEP] [CLS] yeah [MASK] you know and it [MASK] are many of them there'[MASK] no doubt about it [SEP] there [MASK] countless numbers of them, that is for sure. [SEP] [CLS] but i [MASK] [MASK] and my [MASK] like cats so [MASK] haven't reached a real agreement on that yet if we get a place where we can have both it ''

'>>> ll be great but until then [SEP] [MASK] husband and [MASK] can't agree on a pet. [SEP] [CLS] yeah [MASK] me [MASK] [MASK] [MASK] me my [MASK] husband's putting two of us through isn't he lucky [SEP] my husband is lucky [MASK] he is putting [MASK] of us through. [SEP] [CLS] never breaks down [MASK] the [MASK]yrofoam [SEP] never [MASK] the styrofoam [MASK] [SEP] [CLS] i haven't seen that [MASK] i and

In [11]:

downsampled_dataset = lm_datasets.train_test_split(
    test_size=0.1, seed=42
)
downsampled_dataset

DatasetDict({
    train: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 24795
    })
    test: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 2756
    })
})

In [12]:

reload(fn)
trainer = fn.train_mlm_model(model,adapter_name,data_collator,tokenizer, downsampled_dataset['train'],downsampled_dataset['test'])

/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/accelerate/accelerator.py:446: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(


In [13]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/87 [00:00<?, ?it/s]

>>> Perplexity: 23.07


In [14]:
trainer.train()

  0%|          | 0/15500 [00:00<?, ?it/s]

{'loss': 2.3204, 'grad_norm': 1.7871910333633423, 'learning_rate': 9.817333333333334e-05, 'epoch': 1.0}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.9531549215316772, 'eval_runtime': 9.6013, 'eval_samples_per_second': 287.045, 'eval_steps_per_second': 9.061, 'epoch': 1.0}
{'loss': 1.9921, 'grad_norm': 1.5396252870559692, 'learning_rate': 9.301333333333335e-05, 'epoch': 2.0}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.8444288969039917, 'eval_runtime': 10.1472, 'eval_samples_per_second': 271.603, 'eval_steps_per_second': 8.574, 'epoch': 2.0}
{'loss': 1.9109, 'grad_norm': 1.5472346544265747, 'learning_rate': 8.785333333333335e-05, 'epoch': 3.0}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.8013226985931396, 'eval_runtime': 10.1978, 'eval_samples_per_second': 270.255, 'eval_steps_per_second': 8.531, 'epoch': 3.0}
{'loss': 1.8814, 'grad_norm': 1.6406781673431396, 'learning_rate': 8.269333333333334e-05, 'epoch': 3.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.7926344871520996, 'eval_runtime': 10.2069, 'eval_samples_per_second': 270.014, 'eval_steps_per_second': 8.524, 'epoch': 4.0}
{'loss': 1.8453, 'grad_norm': 1.7511218786239624, 'learning_rate': 7.753333333333334e-05, 'epoch': 4.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.7367048263549805, 'eval_runtime': 10.2684, 'eval_samples_per_second': 268.397, 'eval_steps_per_second': 8.473, 'epoch': 5.0}
{'loss': 1.8175, 'grad_norm': 1.6385324001312256, 'learning_rate': 7.237333333333334e-05, 'epoch': 5.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.7697116136550903, 'eval_runtime': 10.2026, 'eval_samples_per_second': 270.128, 'eval_steps_per_second': 8.527, 'epoch': 6.0}
{'loss': 1.7966, 'grad_norm': 1.6851807832717896, 'learning_rate': 6.721333333333334e-05, 'epoch': 6.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.7458832263946533, 'eval_runtime': 10.1885, 'eval_samples_per_second': 270.502, 'eval_steps_per_second': 8.539, 'epoch': 7.0}
{'loss': 1.7807, 'grad_norm': 1.6986666917800903, 'learning_rate': 6.206000000000001e-05, 'epoch': 7.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.7087122201919556, 'eval_runtime': 10.1688, 'eval_samples_per_second': 271.026, 'eval_steps_per_second': 8.556, 'epoch': 8.0}
{'loss': 1.7605, 'grad_norm': 1.8125417232513428, 'learning_rate': 5.69e-05, 'epoch': 8.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.7214252948760986, 'eval_runtime': 10.2036, 'eval_samples_per_second': 270.1, 'eval_steps_per_second': 8.526, 'epoch': 9.0}
{'loss': 1.7484, 'grad_norm': 1.7080365419387817, 'learning_rate': 5.174e-05, 'epoch': 9.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6679506301879883, 'eval_runtime': 10.1452, 'eval_samples_per_second': 271.656, 'eval_steps_per_second': 8.576, 'epoch': 10.0}
{'loss': 1.7416, 'grad_norm': 1.762832760810852, 'learning_rate': 4.6580000000000005e-05, 'epoch': 10.99}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6819788217544556, 'eval_runtime': 10.196, 'eval_samples_per_second': 270.303, 'eval_steps_per_second': 8.533, 'epoch': 11.0}
{'loss': 1.7252, 'grad_norm': 1.6478036642074585, 'learning_rate': 4.142666666666667e-05, 'epoch': 11.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6978157758712769, 'eval_runtime': 10.269, 'eval_samples_per_second': 268.381, 'eval_steps_per_second': 8.472, 'epoch': 12.0}
{'loss': 1.7083, 'grad_norm': 1.7836382389068604, 'learning_rate': 3.626666666666667e-05, 'epoch': 12.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6517997980117798, 'eval_runtime': 10.2378, 'eval_samples_per_second': 269.198, 'eval_steps_per_second': 8.498, 'epoch': 13.0}
{'loss': 1.7127, 'grad_norm': 1.6795012950897217, 'learning_rate': 3.1106666666666665e-05, 'epoch': 13.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6488882303237915, 'eval_runtime': 10.1988, 'eval_samples_per_second': 270.228, 'eval_steps_per_second': 8.53, 'epoch': 14.0}
{'loss': 1.6989, 'grad_norm': 1.6525887250900269, 'learning_rate': 2.594666666666667e-05, 'epoch': 14.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6323537826538086, 'eval_runtime': 10.1859, 'eval_samples_per_second': 270.571, 'eval_steps_per_second': 8.541, 'epoch': 15.0}
{'loss': 1.6948, 'grad_norm': 1.774966835975647, 'learning_rate': 2.078666666666667e-05, 'epoch': 15.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6352702379226685, 'eval_runtime': 10.2131, 'eval_samples_per_second': 269.848, 'eval_steps_per_second': 8.518, 'epoch': 16.0}
{'loss': 1.6838, 'grad_norm': 1.8931164741516113, 'learning_rate': 1.563333333333333e-05, 'epoch': 16.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6591559648513794, 'eval_runtime': 10.2635, 'eval_samples_per_second': 268.523, 'eval_steps_per_second': 8.477, 'epoch': 17.0}
{'loss': 1.6803, 'grad_norm': 1.8276159763336182, 'learning_rate': 1.0473333333333334e-05, 'epoch': 17.98}


  0%|          | 0/87 [00:00<?, ?it/s]

{'eval_loss': 1.6334829330444336, 'eval_runtime': 10.3275, 'eval_samples_per_second': 266.861, 'eval_steps_per_second': 8.424, 'epoch': 18.0}


Overwriting existing adapter 'mlm_union_all_TE'.


{'train_runtime': 3571.2574, 'train_samples_per_second': 138.859, 'train_steps_per_second': 4.34, 'train_loss': 1.8053313615587023, 'epoch': 18.0}


TrainOutput(global_step=13950, training_loss=1.8053313615587023, metrics={'train_runtime': 3571.2574, 'train_samples_per_second': 138.859, 'train_steps_per_second': 4.34, 'train_loss': 1.8053313615587023, 'epoch': 18.0})

In [15]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/87 [00:00<?, ?it/s]

>>> Perplexity: 5.06


In [16]:
import config.config as config

trainer.model.save_adapter(f"{config.Config.ADAPTER_SAVE_PATH}/{adapter_name}", adapter_name)


: 

#### This is a report of mlm for Slate domain using whole genre as target and then splitt after tokenization with seed 42. I used computer metrix for preplexity and early stopping 
The results before:
>>> Perplexity: 14.65

The results after:
>>> Perplexity: 5.06